Develop SearchsortedOp to wrap numpy's searchsorted function

tsirif opened the pull request:
Includes:  
* A `SearchsortedOp`, which includes `perform` and `c_code`
* A `searchsorted` function with numpy-equivalent call
* A method in tensor's basic methods which redirects to `searchsorted` with numpy-equivalent call to ndarray's method
* Unit tests for `SearchsortedOp`'s functionalities

Currently, `side` string argument is being given to SearchSortedOp as an init argument. A param version is not implemented in this patch.

nouiz commented on theano/tensor/basic.py:
```diff
@@ -147,6 +147,8 @@ def as_tensor_variable(x, name=None, ndim=None):
     AsTensorError
         If `x` cannot be converted to a TensorType Variable.
+    TypeError
+        If `x` cannot be made into a Variable with `ndim` dimensions.
```
I get a value error in that case:
```
>>> import theano
>>>v=theano.tensor.matrix()
>>> theano.tensor.as_tensor_variable(v, ndim=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "theano/tensor/basic.py", line 182, in as_tensor_variable
    % ndim, x.type
ValueError: ('TensorType could not be cast to have 1 dimensions', TensorType(float64, matrix))
```

In which case you get that error?

tsirif replied:
In the params doc above, it says:
"""
ndim : None or integer
      Return a Variable with this many dimensions. Raise TypeError if it's
      not possible.
"""

I'm looking into it

nouiz commented on theano/tensor/extra_ops.py:
```diff
@@ -65,6 +65,132 @@ def c_code_cache_version(self):
+        Contains indices that sort array `x` into ascending order.
+        They are typically the result of argsort.
+
+    .. versionadded:: 0.8.2
```
will be added in 0.9

nouiz commented on theano/tensor/extra_ops.py:
```diff
@@ -65,6 +65,132 @@ def c_code_cache_version(self):
+    Returns
+    -------
+    indices : tensor of integers (int64)
+        Array of insertion points with the same shape as `v`.
```
add a link to numpy doc for the examples, or copy the examples.

nouiz replied:
add a note that it only work on CPU.

nouiz commented on theano/tensor/tests/test_extra_ops.py:
```diff
@@ -33,6 +35,123 @@ def test_cpu_contiguous():
+
+    def test_searchsortedOp_on_sorted_input(self):
+        f = theano.function([self.x, self.v], searchsorted(self.x, self.v),
+                            mode="DebugMode")
```
don't put a mode here. In our daily buildbot, we test with many different modes.

nouiz commented on theano/tensor/extra_ops.py:
```diff
@@ -65,6 +65,132 @@ def c_code_cache_version(self):
+        if v.ndim == 1:
+            v_grad = v.zeros_like()
+        else:
+            v_grad = theano.gradient.grad_not_implemented(
```
why ndim > 1 is different? I would think it is the same.

@lamblin can you confirm this method is well implemented? I think so but would like a second opinion.

lamblin replied:
`x_grad` should be `x.zeros_like()` only if `x` has a float-like `dtype`, and `x.zeros_like(floatX)` for int-like `dtype` (gradients have to be float). There should be a helper function doing that somewhere.
Same for `v_grad`, it should be `v.zeros_like(...)` regardless of the number of dimensions.

`sorter_grad` should be `DisconnectedType()` .

tsirif replied:
So, would `x.zeros_like(dtype=floatX)` suffice for both cases? (Same for `v`)

lamblin replied:
Not really, since `x` could still be `float64` when `floatX == 'float32'`, or the other way around. In that case, we really want `x.dtype`, not `floatX`.

nouiz commented on theano/tensor/tests/test_extra_ops.py:
```diff
@@ -33,6 +35,123 @@ def test_cpu_contiguous():
+                            searchsorted(self.x, self.v, sorter=sorter))
+        #  assert np.allclose(
+        #          np.searchsorted(self.a, self.b, sorter=None),
+        #          f(self.a, self.b, sorter=None))
```
Here, you must pass a vector for the sorter input. That is why it raise an error.

nouiz replied:
In fact, this whole test isn't useful. It is the Theano function that will parse the input and will reject the None as it expect a vector. So this don't test this new op. I would just remove this test method.

nouiz commented on theano/tensor/tests/test_extra_ops.py:
```diff
@@ -33,6 +35,123 @@ def test_cpu_contiguous():
+                np.searchsorted(self.a, self.b, side='right'),
+                f(self.a, self.b))
+
+    def test_use_c_code(self):
```
We test the Python and C code in travis and in the daily buildbot. So no need for this test method.

nouiz commented on theano/tensor/tests/test_extra_ops.py:
```diff
@@ -33,6 +35,123 @@ def test_cpu_contiguous():
+                                self.op_class)
+
+        # Test parameter ``side``
+        self.a = np.ones(10).astype(config.floatX)
```
I'm not sure if it is garanty the tests a run in a fixed order. To be safe, don't override this instance attribute in the test.

nouiz commented:
This PR look pretty good. I just got a few questions/comments.

thanks

tsirif commented:
Thank you Fred. I am working on these comments right now. 2 questions:
1. does `side` argument need to be a dynamic op param or not?  
2. About ndim > 1 in grad method: Documentation refers that handling grads of non vector inputs is not implemented yet.

nouiz commented:
side don't need to be a dinamic param. But it can be. You can wrap the
string in a Generic() variable.

The other option is to use params.

Both change would allow to compile less version of this code. Otherwise, we
need to compile 2 versions of the code.

This isn't a must do to merge this PR, but if we could compile just one
version, it would be great.

2) I'm not sure what you are talking about. Here the grad implementation
would be the same as for 1d, zeros(). Where did you saw this?

On Fri, Apr 22, 2016 at 3:35 PM, Christos Tsirigotis <
notifications@github.com> wrote:

> Thank you Fred. I am working on these comments right now. 2 questions:
> 1. does side argument need to be a dynamic op param or not?
>
> 2. About ndim > 1 in grad method: Documentation refers that handling grads
> of non vector inputs is not implemented yet.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4422#issuecomment-213563350>
>

tsirif commented:
1) Don't merge yet, I will fix what you commented and also I will make `side` a dynamic param for practice.

2) I cannot find it right now. Maybe it was a misconception of mine. I will fix this.

3) Thanks for accepting me in GSoC! :smiley:

lamblin commented on theano/tensor/extra_ops.py:
```diff
@@ -65,6 +65,132 @@ def c_code_cache_version(self):
+def searchsorted(x, v, side='left', sorter=None):
+    """Find indices where elements should be inserted to maintain order.
+
+    Wraping of numpy.searchsorted. Find the indices into a sorted array
```
"Wrapping"

tsirif replied:
oups.

lamblin commented on theano/tensor/extra_ops.py:
```diff
@@ -65,6 +65,132 @@ def c_code_cache_version(self):
+    x: 1-D tensor (array-like)
+        Input array. If `sorter` is None, then it must be sorted in
+        ascending order, otherwise `sorter` must be an array of indices
+        that sort it.
```
"sorts"

tsirif commented:
Ok, I think I completed changes according to your comments and I also added `side` as a dynamic param.

lamblin commented:
General note: could you rebase your changes on top of the master branch, rather than merging with it? Merging brought in unrelated changes in this PR.
Also, on test is failing, it looks like numerical gradient estimation went across a point where the gradient is not defined, causing it to expect a huge gradient.

tsirif commented:
@lamblin, what should be done on the failing test?

tsirif commented:
Also, you mean that I should not merge with no-ff into the branch on which I will issue a pull request, right? You do not want e.g. `Merge branch 'feature/searchsorted' into develop` commits into history tree. I commit frequently though small partial changes which may not be complete code features to be run through ci yet and I do not want to spam hooks on travis with every push. Should I not care about this? Should I use `[ci skip]` in commit messages or do you suggest something else?

lamblin commented:
For the failing test, I would suggest sampling fewer values, or spanning a larger range, to avoid almost-collisions in the values.
For the merge, that is right, you should avoid *merging* with `central/master` or `upstream/master` while developing. If there is a conflict or that you need to update, it is better to use `git rebase -i central/master` instead. This will "lift" your commits on top of the latest master commits and apply them there.
You can have partial local commits and push once in a while, I've never used `[ci skip]` but it may be a good idea.

tsirif commented:
@lamblin : So far I used to have a `feature/<something>` branch in which I was working and when then the next x commits in this branch had completed a code feature, I was merging a `develop` branch with `--no-ff` to `feature/<something>`. `develop` branch is the one that I issue for a pull request, so every time a new merging commit is pushed to the repo, a new necessary hook on travis begins. If I keep rebasing this subtree on top of most recent `theano/master` branch by the time I want to push on `develop`, would this be a desirable development scheme?

tsirif commented:
I rebased on top of current `theano/master` and squashed commits into a single commit for every merge I had done in my `develop` from `feature/searchsorted` branch. The build is passing successfully.

nouiz commented:
You can rebase on top of the central/master when needed. But you don't need to always do it. Most of the time it isn't needed.

But when a rebase/merge is needed, make a rebase as you did here. I'll check again this PR.

nouiz commented on theano/tensor/extra_ops.py:
```diff
@@ -68,6 +70,164 @@ def c_code_cache_version(self):
+    __props__ = ("side", )
+
+    def __init__(self, side='left'):
+        self.side = side
```
Add an assert that side is "left" or "right".

tsirif replied:
Maybe I should let numpy's exception to be raised. I will change the c_code implementation to allow this.

nouiz replied:
It is better to return the error when we build the graph when possible then during execution. So I would add the check here.

nouiz commented on theano/tensor/tests/test_extra_ops.py:
```diff
@@ -37,6 +38,80 @@ def test_cpu_contiguous():
+        self.b = 30 * np.random.random((8, 10, 5)).astype(config.floatX)
+        self.idx_sorted = np.argsort(self.a)
+
+    def tearDown(self):
```
I don't understand why this is needed. We don't do it elsewhere. Is there a particular reason you do this? I think by default all member get removed and Python gc would collect them afterward. But if you know they are needed, please explain. I never check my hypothesis.

tsirif replied:
You are right, this is redundant. I will remove it.

nouiz commented on theano/tensor/var.py:
```diff
@@ -692,6 +692,9 @@ def cumsum(self, axis=None):
     def cumprod(self, axis=None):
         return theano.tensor.extra_ops.cumprod(self, axis)
 
+    def searchsorted(self, v, side='left', sorter=None):
```
This isn't used in your test. Just change one test to call it by this way. This make sure we won't break/remove that interface by mistake.

tsirif replied:
:+1:

nouiz commented:
I got very small/comments questions, so we should be able to merge this soon.

thanks.

tsirif commented:
@nouiz I pushed fixes

MarcCote commented:
@tsirif well done. :+1:
